pkg/tests/cross-tests/tfwrite.go

```diff
@@ -39,42 +39,83 @@ func WriteHCL(out io.Writer, sch map[string]*schema.Schema, resourceType, resour
 }
 
 func writeBlock(body *hclwrite.Body, schemas map[string]*schema.Schema, values map[string]cty.Value) {
-	keys := make([]string, 0, len(schemas))
-	for key := range schemas {
-		keys = append(keys, key)
+	internalMap := schema.InternalMap(schemas)
+	coreConfigSchema := internalMap.CoreConfigSchema()
+
+	blockKeys := make([]string, 0, len(coreConfigSchema.BlockTypes))
+	for key := range coreConfigSchema.BlockTypes {
+		blockKeys = append(blockKeys, key)
 	}
-	sort.Strings(keys)
-	for _, key := range keys {
-		sch := schemas[key]
-		value, ok := values[key]
-		if !ok {
-			continue
-		}
+	sort.Strings(blockKeys)
 
-		switch elem := sch.Elem.(type) {
-		case *schema.Resource:
-			if sch.Type == schema.TypeMap || sch.ConfigMode == schema.SchemaConfigModeAttr {
-				body.SetAttributeValue(key, value)
-			} else if sch.Type == schema.TypeSet {
-				for _, v := range value.AsValueSet().Values() {
-					newBlock := body.AppendNewBlock(key, nil)
-					writeBlock(newBlock.Body(), elem.Schema, v.AsValueMap())
-				}
-			} else if sch.Type == schema.TypeList {
-				for _, v := range value.AsValueSlice() {
-					newBlock := body.AppendNewBlock(key, nil)
-					writeBlock(newBlock.Body(), elem.Schema, v.AsValueMap())
-				}
-			} else {
-				contract.Failf("unexpected schema type %v", sch.Type)
+	for _, key := range blockKeys {
+		bl := coreConfigSchema.BlockTypes[key]
+		switch bl.Nesting.String() {
+		case "NestingSingle":
+			v, ok := values[key]
+			if !ok {
+				continue
+			}
+			newBlock := body.AppendNewBlock(key, nil)
+			res, ok := schemas[key].Elem.(*schema.Resource)
+			if !ok {
+				contract.Failf("unexpected schema type %s", key)
+			}
+			writeBlock(newBlock.Body(), res.Schema, v.AsValueMap())
+		case "NestingGroup":
```

Comment:
INteresting, this one I'm not that familiar with, you have a test for it I bet? Let me look.